### ISSUE TITLE
Puts is required to get proper data when a variable is passed through a proc

### ISSUE THREAD

**Issue by venomc312** [crystal-lang/crystal#10911]
## Bug Report

# Please read

```
As I do not know the specifics to this bug I would like to link my repository.  I'm going to make it temporarily public so it can be reviewed and sorted.  The link will follow.  Also attached to this bug report is the issue that was created under said repository.
```

[Said Repository](https://github.com/venomc312/vcbot)

When creating a command in a separate file in "./commands/"; you will not receive the right payload data through the proc variable. The data coming through is somehow messed up through the type inference or something of the likes. This is not a program issue and rather a language or discordcr issue.

In order to fix this you must print one of the functions that use the payload data, this somehow makes the data proper. Doing this will ensure that you receive the right data. The only downside to this temporary fix is that the console will be filled with useless data.

**Comment by straight-shoota** [comment]
Please provide specific information about the alleged bug: the code in question as well as expected and actual behaviour and clear instructions how to reproduce. We don't know your code and how to run it in order to figure that out.

Missing a clear bug description, this seems more like a general request for help which is better served at https://forum.crystal-lang.org/

**Comment by z64** [comment]
(Context: I am one of the authors of the library being used here)

This is the code in question:

```cr
module Commands
    extend self
    def ping(payload : Discord::Message)
        puts Discord.client.create_message(payload.channel_id.value, "Pong!")
    end
    Discord.commands["ping"] = ->ping(Discord::Message)
end
```
https://github.com/venomc312/vcbot/blob/master/vcbot/src/commands/ping.cr#L8

I stepped through this with `gdb`. What happens is, if the `puts` is not present, it generates different "inlined" assembly that causes the `payload.channel_id` field of the `Discord::Message` struct to be overwritten with a value from the stack. There is no Crystal code that does this (the object is immutable). In turn, when `create_message` is called, it tries to send a message to a bogus channel ID.

Some facts / observations:
- `channel_id` is a UInt64 that is passed to `create_message` via a register. They are always large numbers, example: `653812356656398380`
- The value it is overwritten with is some low value, it varies, but for example `0x50b`. I have been unable to pinpoint where this value comes from. It is pulled from the stack somewhere from a parent frame.
- The difference that `puts` makes causes the values to be allocated / passed through different registers, causing the value to not be overwritten.

---

The user takes the approach of assigning a proc in a hash map to dispatch commands:

```cr
Discord.commands["ping"] = ->ping(Discord::Message)
```

The dispatch happens here, in a callback provided by the library:
https://github.com/venomc312/vcbot/blob/master/vcbot/src/discord.cr#L35

If you replace this dispatch through the hash map with a direct method call, i.e. 

```diff
- Discord.commands[cmdname].call(payload)
+ Commands.ping(payload)
```

It works fine.

So, the issue seems to be something involving the calling procedure of procs via this hash map. But I've been unable to reduce it thus far. I will try again later.

**Comment by z64** [comment]
@venomc312 As far as I can tell thus far, Crystal is generating some bad code for this mechanism you've written. As workaround, please take an alternate approach. There are many Discord bots to take inspiration from with command systems that you can find in GitHub, or find suggestions in the official support channel for `discordcr`.

**Comment by venomc312** [comment]
Yes, this is why I opened the issue here, when being called normally through a direct method call it works fine, but when passed through proc it does not.  The expected value is supposed to be UInt64 while what I'm getting is not.  Thank you for commenting by the way.  I'd like to help get this bug solved :)

**Comment by venomc312** [comment]
As the data being passed through proc is only correct when using a `puts` statement; I suspect that it is a bug with the language hence why I opened the report here :)

**Comment by venomc312** [comment]
z64 has a much better explanation, as i'm newer to this language I didn't know there was even a way to walk through the code.  I will be using that in the future.  I could also provide more information if needed.

**Comment by z64** [comment]
Ok, I am *fairly* certain this is an accurate reduction.

```cr
struct Data
  getter int : UInt64

  getter _pad1 : UInt64?

  # NOTE(z64): removing this line fixes the issue
  getter _pad2 : UInt64?

  def initialize(@int : UInt64)
  end
end

def exec(int : UInt64)
  raise "BUG: #{int}" if int != 653812356656398380_u64

  # NOTE(z64): removing this line fixes the issue
  Data.new(0)
end

# NOTE(z64): replacing this with *just* a proc fixes the issue
commands = Hash(String, Proc(Data, Nil)).new

commands["foo"] = ->(payload : Data) do
  exec(payload.int)
end

data = Data.new(653812356656398380_u64)
commands["foo"].call(data)
```

This crashes with:
```
Unhandled exception: BUG: 0 (Exception)
  from src/discord.cr:14:3 in 'exec'
  from src/discord.cr:23:3 in '->'
  from /usr/lib/crystal/primitives.cr:255:3 in '__crystal_main'
  from /usr/lib/crystal/crystal/main.cr:110:5 in 'main_user_code'
  from /usr/lib/crystal/crystal/main.cr:96:7 in 'main'
  from /usr/lib/crystal/crystal/main.cr:119:3 in 'main'
  from __libc_start_main
  from _start
  from ???
```

While I was reducing this, I was able to get various `BUG` values, including garbage, values of adjacent fields, and at one point a segfault.

**Comment by asterite** [comment]
Further reduced:

```crystal
casted_proc : Proc(Int32, Nil) = ->(int : Int32) do
  p int
  {10_i64, 20_i64, 30_i64, 40_i64}
end

casted_proc.call(1)
```

We allow implicit casting from a Proc that doesn't return `Nil` to a Proc that does return `Nil`, but I think if the return value is bigger than some bytes then it doesn't work well. I still don't understand what the conditions are exactly. This also reproduces it:

```crystal
casted_proc : Proc(Int32, Nil) = ->(int : Int32) do
  p int
  {1, 2, 3, 4, 5}
end

casted_proc.call(1)
```

And this:

```crystal
casted_proc : Proc(Int32, Nil) = ->(int : Int32) do
  x = uninitialized StaticArray(UInt8, 4)
  x
end

casted_proc.call(1)
```

Maybe the proc is called, some value is put on the stack, and we don't pop it and that's wrong. For values less than X bytes the value is put on a register? No idea.

In any case, we probably need to change the language so that this cast is invalid if the return value is a struct type. That's a breaking change, but it's more like a bug fix.

The workaround is to put `nil` at the end of the proc literal, or just use an alias:

```crystal
alias Command = Proc(Data, Nil)

commands = Hash(String, Command).new

commands["foo"] = Command.new do |payload|
  exec(payload.int)
end
```

**Comment by asterite** [comment]
Oops: https://stackoverflow.com/a/47089563/4990850

I think we need to fix this to stop compiling. Maybe we can allow this to work for primitives (ints, floats, etc.) and reference types, but nothing else. Or maybe we can create a wrapper function, but that's an extra closure so I'm not sure.

**Comment by venomc312** [comment]
The expected value for what is in the ping command is a UInt64, the payload is parsed as a json through a packet in the library it's really just a json object I believe but I'm not totally sure.  If @z64 could iterate on this that would be great.

**Comment by asterite** [comment]
Thank you!

The bug is super clear now, we just need to decide how and when to fix it. No further detail are needed.

**Comment by z64** [comment]
@venomc312 The problem isn't with the message payload from the event - it is with the return type of `create_message`, which returns the created message. 

By writing `puts Discord.client.create_message(...)` you incidentally changed the return type of `ping` from `Discord::Message` to `nil`, which fixed the issue at hand. The workaround as mentioned is to explicitly return `nil` at the end of your `ping` method.

So, going back to your original code, this would be a succinct way to fix it if you want to keep using that pattern:

```diff
  module Commands
    extend self
-   def ping(payload : Discord::Message)
+   def ping(payload : Discord::Message) : Nil
      Discord.client.create_message(payload.channel_id.value, "Pong!")
    end
    Discord.commands["ping"] = ->ping(Discord::Message)
  end
```

The return type restriction here forces `ping` to implicitly return `nil`.

**Comment by venomc312** [comment]
> @venomc312 The problem isn't with the message payload from the event - it is with the return type of `create_message`, which returns the created message. 
> 
> By writing `puts Discord.client.create_message(...)` you incidentally changed the return type of `ping` from `Discord::Message` to `nil`, which fixed the issue at hand. The workaround as mentioned is to explicitly return `nil` at the end of your `ping` method.
> 
> So, going back to your original code, this would be a succinct way to fix it if you want to keep using that pattern:
> 
> ```diff
>   module Commands
>     extend self
> -   def ping(payload : Discord::Message)
> +   def ping(payload : Discord::Message) : Nil
>       Discord.client.create_message(payload.channel_id.value, "Pong!")
>     end
>     Discord.commands["ping"] = ->ping(Discord::Message)
>   end
> ```
> 
> The return type restriction here forces `ping` to implicitly return `nil`.
> 
> And to be doubly clear, this is not a problem with `def` in general, but the fact that you are using method procs, `->ping(msg : Discord::Message)`

Thank you, this will streamline the process for now until this bug is fixed :)

**Comment by venomc312** [comment]
Could you all possibly let me know when I can set my repo back to private, Thanks :)

**Comment by z64** [comment]
@venomc312 We have a minimal reproduction now that doesn't require your code. It is fine if you want to remove it.

**Comment by venomc312** [comment]
@z64 Thank you!

**Comment by HertzDevil** [comment]
It seems the `Proc`s that return large structs do expect the stack to be set up by the caller: https://godbolt.org/z/E5qvP6GPb

I think it is okay to create a wrapper closure here; we already do that for some kinds of proc pointers. So we would have:

```crystal
class Foo
  def bar
    1
  end
end

foo = Foo.new
f : Proc(Nil) = ->foo.bar

# the above becomes:

f : -> Nil = begin
  __temp_1 = begin
    __temp_2 = foo
    -> { __temp_2.bar }
  end
  -> { __temp_1.call; nil }
end
```

**Comment by asterite** [comment]
@HertzDevil Yes, but note that this type of casting is done in a lot of places. So for example if you do this:

```crystal
proc = ->(x : Int32, y : Int32) { {x, y, 3, 4} }
f : Proc(Int32, Int32, Nil) = proc
```

We would probably have to rewrite a proc being assigned like that to this:

```crystal
proc = ->(x : Int32, y : Int32) { {x, y, 3, 4} }
f : Proc(Int32, Int32, Nil) = ->(x : Int32, y : Int32) { proc.call(x, y); nil }
```

But introducing a hidden closure isn't really nice... but maybe it's not a big deal, I don't know. I think ideally we should simply error and ask the user to explicitly return `nil`, but maybe it's a huge breaking change at this point so the closure solution would be fine.

**Comment by asterite** [comment]
@straight-shoota This is the implicit Proc conversion I'm talking about.

I think we should eventually fix it by either completely disallowing it, or by introducing a wrapper struct (but it creates a hidden closure and a hidden allocation, so I personally don't like this solution.)

**Comment by asterite** [comment]
We could define the wrapper proc in Crystal itself, maybe put it in `prelude.cr`:

```crystal
def __crystal_wrap_proc(proc : Proc(*T, R)) : Proc(*T, Nil) forall T, R
  {% begin %}
  ->(
      {% for t, i in T %}
        arg{{i}} : {{t}},
      {% end %}
    )
  {
    proc.call(
      {% for t, i in T %}
        arg{{i}},
      {% end %}
    )
    nil
  }
  {% end %}
end
```

Then whenever the compiler or the interpreter needs to convert a Proc(*T, R) to Proc(*T, Nil), we just do a call to `__crystal_wrap_proc` with that proc (though we'd have to type it, but it will always succeed).

**Comment by straight-shoota** [comment]
What's the benefit of defining this wrapper in the prelude vs. implementing it directly in the compiler?

Alternative idea: Could we perhaps inline this at the callsite? So that a call to `proc.call` when the variable `proc` is restricted to `Proc(*X, Nil)` but the actual type is `Proc(*X, R)`, the call gets replaced by `(proc.call; nil)`.

**Comment by asterite** [comment]
> Could we perhaps inline this at the callsite?

No, because `proc.call` already produces wrong code if the return type is not correct. For the interpreter, `proc.call` will call the proc and the pop 0 bytes, but the actual proc put some return value in the stack. For compiled mode, I think the proc args are already wrong (or I'd have to read this thread to understand what's going on)

I'm actually thinking that a wrapper struct is not good. For example if a C function expects a proc that returns `Nil`, but we pass it something that doesn't return `Nil`, we'll create the wrapper. But then it's a closure! So it won't compile. But the user will be very confused about the error message.

So here's what I think we should do:
- If there's a conversion between `Proc(*R, T)` and `Proc(*R, Nil)` and the arg is a Proc literal, we can rewrite the literal to return `nil`.
- If it's not a proc literal, we just error.

I believe in most cases this will be backwards compatible. In other cases it will produce a compile error, but maybe we can check the ecosystem specs and fix what's needed. If it's a bug in the language and it implies a breaking change, I think we should still move forward with it.

**Comment by asterite** [comment]
Interestingly, I can't seem to reproduce the original issue now. [This code snippet](https://github.com/crystal-lang/crystal/issues/10911#issuecomment-877163793) doesn't crash for me.

**Comment by asterite** [comment]
But, I think we should still move forward with this.

Here's an update to what I plan to do:
- If there's a conversion between Proc(*R, T) and Proc(*R, Nil) and the arg is a Proc literal, we can rewrite the literal to return nil
- If it's not a proc literal, we produce a warning saying that this should be updated instead to return `nil`, and that eventually this warning will be an error, so it's better if it's fixed soon :-)

That way we don't actually break anyone's code.

What do you think?

**Comment by z64** [comment]
> Interestingly, I can't seem to reproduce the original issue now. [This code snippet](https://github.com/crystal-lang/crystal/issues/10911#issuecomment-877163793) doesn't crash for me.

I can still repro on 1.5:

```
Unhandled exception: BUG: 0 (Exception)
  from test3.cr:14:3 in 'exec'
  from test3.cr:24:3 in '->'
  from test3.cr:28:17 in '__crystal_main'
  from /usr/lib/crystal/crystal/main.cr:115:5 in 'main_user_code'
  from /usr/lib/crystal/crystal/main.cr:101:7 in 'main'
  from /usr/lib/crystal/crystal/main.cr:127:3 in 'main'
  from /usr/lib/libc.so.6 in '??'
  from /usr/lib/libc.so.6 in '__libc_start_main'
  from ../sysdeps/x86_64/start.S:117 in '_start'
  from ???
```

```
Crystal 1.5.0 (2022-07-21)

LLVM: 14.0.6
Default target: x86_64-pc-linux-gnu
```

**Comment by asterite** [comment]
Thanks! I guess it just doesn't crash on Mac.

**Comment by zw963** [comment]
Hope this issue can be fixed ASAP, because as discussed in [this thread](https://forum.crystal-lang.org/t/invalid-memory-access-raised-instead-of-expected-cast-failed-error-what-tools-should-i-use-to-debug-this-type-of-error/5198/7), it caused Invalid Memory Access issue, which really hard to figure out.

**Comment by HertzDevil** [comment]
Note: instead of adding `nil` to the `ProcLiteral` body, it is now possible to add or overwrite the literal's return type with `Nil`

**Comment by zw963** [comment]
> Note: instead of adding `nil` to the `ProcLiteral` body, it is now possible to add or overwrite the literal's return type with `Nil`

I guess what @HertzDevil  means is. 

instead of following code broken with segment fault.

```crystal
casted_proc : Proc(Int32, Nil) = ->(int : Int32) do
  p int
  {1, 2, 3, 4, 5}
end

casted_proc.call(1)
```

```
Invalid memory access (signal 11) at address 0x11
[0x5c44b5bac726] *Exception::CallStack::print_backtrace:Nil +118 in /home/zw963/.cache/crystal/crystal-run-tmp.Rex6fiFHuC.tmp
[0x5c44b5b6a6f6] ~procProc(Int32, Pointer(LibC::SiginfoT), Pointer(Void), Nil) +310 in /home/zw963/.cache/crystal/crystal-run-tmp.Rex6fiFHuC.tmp
[0x7d6308435770] ?? +137864293865328 in /usr/lib/libc.so.6
[0x5c44b5b6d22e] ~procProc(Int32, Tuple(Int32, Int32, Int32, Int32, Int32)) +4078 in /home/zw963/.cache/crystal/crystal-run-tmp.Rex6fiFHuC.tmp
[0x5c44b5b54d9b] __crystal_main +1595 in /home/zw963/.cache/crystal/crystal-run-tmp.Rex6fiFHuC.tmp
[0x5c44b5c7d066] *Crystal::main_user_code<Int32, Pointer(Pointer(UInt8))>:Nil +6 in /home/zw963/.cache/crystal/crystal-run-tmp.Rex6fiFHuC.tmp
[0x5c44b5c7cfda] *Crystal::main<Int32, Pointer(Pointer(UInt8))>:Int32 +58 in /home/zw963/.cache/crystal/crystal-run-tmp.Rex6fiFHuC.tmp
[0x5c44b5b64b06] main +6 in /home/zw963/.cache/crystal/crystal-run-tmp.Rex6fiFHuC.tmp
[0x7d630841ecd0] ?? +137864293772496 in /usr/lib/libc.so.6
[0x7d630841ed8a] __libc_start_main +138 in /usr/lib/libc.so.6
[0x5c44b5b54685] _start +37 in /home/zw963/.cache/crystal/crystal-run-tmp.Rex6fiFHuC.tmp
[0x0] ???
```
------------------


The old (when this issue was created) solution is:

```crystal
casted_proc : Proc(Int32, Nil) = ->(int : Int32) do
  p int
  {1, 2, 3, 4, 5}
  nil
end

casted_proc.call(1)
```

-------------------------

We can fix it now like this:

```
casted_proc : Proc(Int32, Nil) = ->(int : Int32) : Nil do
  p int
  {1, 2, 3, 4, 5}
end

casted_proc.call(1)
```

Because I have encountered this problem before, when segment fault occurs, I will try figure it out following this issue.
But, for developers with less experience, this issue that causes Invalid memory access (there is no way to debug at all) is enough to him give up Crystal.

Anyway. above just a workaround, never fix it, i really don't know why this fix is breaking changes(it cause a Invalid memory access, right), and we can't fix it with completely disallowing it? can someone explain this? thanks.
